src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -222,51 +223,242 @@ private bool ScanDesignation(bool permitTuple)
+                        subPatterns[0].Pattern is ConstantPatternSyntax cp)
+                    {
+                        // There is an ambiguity between a deconstruction pattern `(` pattern `)`
+                        // and a constant expression pattern than happens to be parenthesized.
+                        // We normalize to the parenthesized expression, and semantic analysis
+                        // will notice the parens and treat it whichever way is semantically sensible,
+                        // giving priority to its treatment as a constant expression.
+                        return _syntaxFactory.ParenthesizedExpression(openParenToken, cp.Expression, closeParenToken);
                     }
+
+                    var node = _syntaxFactory.DeconstructionPattern(type, openParenToken, subPatterns, closeParenToken, propertySubpattern, designation);
+                    return this.CheckFeatureAvailability(node, MessageID.IDS_FeatureRecursivePatterns);
                 }
-                finally
+
+                if (parsePropertySubpattern())
                 {
-                    this.Release(ref resetPoint);
+                    parseDesignation();
+                    var node = _syntaxFactory.PropertyPattern(type, propertySubpattern, designation);
+                    return this.CheckFeatureAvailability(node, MessageID.IDS_FeatureRecursivePatterns);
+                }
+
+                if (type != null && parseDesignation())
+                {
+                    return _syntaxFactory.DeclarationPattern(type, designation);
                 }
+
+                this.Reset(ref resetPoint);
+                return this.ParseSubExpression(Precedence.Expression);
             }
-            else
+            finally
             {
-                // In places where a pattern is supported, we do not support tuple types
-                // due to both syntactic and semantic ambiguities between tuple types and positional patterns.
+                this.Release(ref resetPoint);
+            }
+        }
 
-                // But it still might be a pattern such as (operand is 3) or (operand is nameof(x))
-                node = this.ParseSubExpression(Precedence.Expression);
+        private PatternSyntax ParsePattern()
+        {
+            var node = ParseExpressionOrPattern(forCase: false);
+            switch (node)
+            {
+                case PatternSyntax pattern:
+                    return pattern;
+                case ExpressionSyntax expression:
+                    return _syntaxFactory.ConstantPattern(expression);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(node);
             }
+        }
+
+        private PropertySubpatternSyntax ParsePropertySubpattern()
+        {
+            ParseSubpatternList(out var openBraceToken, out var subPatterns, out var closeBraceToken, SyntaxKind.OpenBraceToken, SyntaxKind.CloseBraceToken);
```

Comment:
>ParseSubpatternList(out var openBraceToken, out var subPatterns, out var closeBraceToken, SyntaxKind.OpenBraceToken, SyntaxKind.CloseBraceToken); [](start = 12, length = 145)

Please spell out types.

Reply:
[https://github.com/dotnet/roslyn/issues/23222](https://github.com/dotnet/roslyn/issues/23222) already has an issue to spell out type names in out variable declarations.

---
In reply to: [151815236](https://github.com/dotnet/roslyn/pull/23208#discussion_r151815236) [](ancestors = 151815236)